plugin/acl/setup.go

```diff
@@ -0,0 +1,181 @@
+package acl
+
+import (
+	"net"
+	"strings"
+
+	"github.com/coredns/coredns/core/dnsserver"
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/metrics"
+
+	"github.com/caddyserver/caddy"
+	"github.com/infobloxopen/go-trees/iptree"
+	"github.com/miekg/dns"
+)
+
+const (
+	// QtypeAll is used to match any kinds of DNS records type.
+	// NOTE: The value of QtypeAll should be different with other QTYPEs defined in miekg/dns.
+	QtypeAll uint16 = 0
+)
+
+func init() {
+	caddy.RegisterPlugin("acl", caddy.Plugin{
+		ServerType: "dns",
+		Action:     setup,
+	})
+}
+
+func createDefaultFilter() *iptree.Tree {
+	defaultFilter := iptree.NewTree()
+	_, IPv4All, _ := net.ParseCIDR("0.0.0.0/0")
+	_, IPv6All, _ := net.ParseCIDR("::/0")
+	defaultFilter.InplaceInsertNet(IPv4All, struct{}{})
+	defaultFilter.InplaceInsertNet(IPv6All, struct{}{})
+	return defaultFilter
+}
+
+func setup(c *caddy.Controller) error {
+	a, err := parse(c)
+	if err != nil {
+		return plugin.Error("acl", err)
+	}
+
+	dnsserver.GetConfig(c).AddPlugin(func(next plugin.Handler) plugin.Handler {
+		a.Next = next
+		return a
+	})
+
+	// Register all metrics.
+	c.OnStartup(func() error {
+		metrics.MustRegister(c, RequestBlockCount, RequestAllowCount)
+		return nil
+	})
+	return nil
+}
+
+func parse(c *caddy.Controller) (acl, error) {
+	a := acl{}
+	for c.Next() {
+		r := Rule{}
+		r.Zones = c.RemainingArgs()
+		if len(r.Zones) == 0 {
+			// if empty, the zones from the configuration block are used.
+			r.Zones = make([]string, len(c.ServerBlockKeys))
+			copy(r.Zones, c.ServerBlockKeys)
+		}
+		for i := range r.Zones {
+			r.Zones[i] = plugin.Host(r.Zones[i]).Normalize()
+		}
+
+		for c.NextBlock() {
+			p := Policy{}
+
+			action := strings.ToLower(c.Val())
+			if action == "allow" {
+				p.action = Allow
+			} else if action == "block" {
+				p.action = Block
+			} else {
+				return a, c.Errf("unexpected token %q; expect 'allow' or 'block'", c.Val())
+			}
+
+			p.qtypes = make(map[uint16]struct{})
+
+			p.filter = iptree.NewTree()
```

Comment:
why doesn't this use createDefaultFilter?

Reply:
default filter includes the widest possible IPv4 and IPv6 networks, i.e. any IP hits this filter.
Here we need to start with empty filter, i.e. no IP hits this filter initially. Upcoming lines will add the configured networks

Reply:
`newDefaultFilter` matches all IP addresses while `iptree.NewTree` is an empty filter matches nothing.